### PR TITLE
blockchain/genesis: refactor ChainConfig

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes
Klaytn is preparing for its next hard fork. What matters in it is the promise of the update deadline. Nodes will update the release containing the incompatible changes that cause the hardfork until the deadline. Geth promises this period by setting the hard fork block number in the code.
Since Geth and Klaytn are both open source, private networks are operated in addition to the mainnet/testnet. In particular, using Klaytn can also operate a service chain, one of the layer2 solutions. The characteristic of these chains is that there is a separate node operating entity, and they need to be familiar with hard forks, find related guides, and set up hard fork block numbers.
However, since code related to Klaytn's block number is a little different from geth, in order to prevent any confusion in advance, we try to match this codebase as much as possible. Therefore this PR is suggested.

* This PR changes mainnet(cypress)/baobab adds chain config location to `params/config.go`
* It also adds incompatible1Block field to the chainConfig struct. The name of the field is temporary.

https://github.com/klaytn/klaytn/pull/882/commits/ca42a114813abbe9d6a893b284caa32c58133ad5: applied due to goimport `consensus/istanbul/validator/default_test.go`

**Test that the compatibility is not broken**

step1. deploy below contract before change binary. It is to check the previously deployed contract is working well.
```
contract Contract {
    function callVmLogPrecompiledContract() public returns (bool) {
        address vmlogaddr = address(0x09);
        vmlogaddr.call("Hello, Klaytn");
    }
}
```
step2. run init. `$ ken init --datadir data genesis.json`

**Chain Configuration Result**
dev
```
INFO ... Initialised chain configuration  config="{ChainID: 940625 Engine: istanbul ...}"
```
PR
```
INFO ... Initialised chain configuration  config="{ChainID: 940625 Incompatible1Block: <nil> ... Engine: istanbul}"
```
**call contract result**(ex. data/ken/logs/vm.log)
dev
```
tx=0x0000000000000000000000000000000000000000000000000000000000000000 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x0000000000000000000000000000000000000000000000000000000000000000 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x474a622c39aa5f230e43dcd2ffa19aec26ccdb915a9cadf21a1e3a39d928dc67 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x474a622c39aa5f230e43dcd2ffa19aec26ccdb915a9cadf21a1e3a39d928dc67 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
```
PR
```
tx=0x0000000000000000000000000000000000000000000000000000000000000000 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x0000000000000000000000000000000000000000000000000000000000000000 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x6b8892220c581eca54278ce5fb1e98322842a2c02b1e313bca1ed63905f35bc0 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x6b8892220c581eca54278ce5fb1e98322842a2c02b1e313bca1ed63905f35bc0 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
```
## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by ehnuje** [comment]
```
consensus/istanbul/validator/default_test.go:32: File is not `goimports`-ed (goimports)
	"testing"
```

@yoomee1313 Linter complains ;)

File: params/config.go [hunk 1]
```diff
@@ -31,21 +31,62 @@ import (
 // Genesis hashes to enforce below configs on.
 var (
 	MainnetGenesisHash      = common.HexToHash("// todo generate new hash for mainnet") // Mainnet genesis hash to enforce below configs on
```
Comment (ehnuje) [inline params/config.go:33]:
```suggestion
	CypressGenesisHash      = common.HexToHash("// todo generate new hash for mainnet") // Mainnet genesis hash to enforce below configs on
```
For alignment with `BaobabGenesisHash`?

**Review by ehnuje — APPROVED** [review]
LGTM except for a minor comment.

File: params/config.go (hunk 1 above)
Comment (yoomee1313) [reply params/config.go]: thanks, reflected at https://github.com/klaytn/klaytn/pull/882/commits/16a16c3814a573dbc65e3c774b753bd33d918755

**Review by ehnuje — APPROVED** [review]

File: params/config.go [hunk 2]
```diff
@@ -30,22 +30,63 @@ import (
...
 )
 
 var (
-	// MainnetChainConfig is the chain parameters to run a node on the main network.
-	MainnetChainConfig = &ChainConfig{
-		ChainID: big.NewInt(1),
-		Gxhash:  new(GxhashConfig),
+	// CypressChainConfig is the chain parameters to run a node on the cypress main network.
+	CypressChainConfig = &ChainConfig{
+		ChainID:            big.NewInt(8217),
```
Comment (winnie-byun) [inline params/config.go:42]: Could you use `CypressNetworkId` instead of `8217`?

File: params/config.go [hunk 3]
```diff
@@ -30,22 +30,63 @@ import (
...
+		UnitPrice: 25000000000,
 	}
 
-	// TestnetChainConfig contains the chain parameters to run a node on the Ropsten test network.
-	TestnetChainConfig = &ChainConfig{
-		ChainID: big.NewInt(2),
-		Gxhash:  new(GxhashConfig),
+	// BaobabChainConfig contains the chain parameters to run a node on the Baobab test network.
+	BaobabChainConfig = &ChainConfig{
+		ChainID:            big.NewInt(1001),
```
Comment (winnie-byun) [inline params/config.go:68]: Could you use `BaobabNetworkId` instead of`1001`?

**Review by winnie-byun — APPROVED** [review]

File: blockchain/genesis_alloc.go [hunk 4]
```diff
@@ -31,29 +31,6 @@ const testnetAllocData = "\xda\u0654\x19t\x18\x1a?\x1bE+\x00\xec\u06ab\x19\xa2j\
 
 var (
 	baobabGenesisJson = []byte(`{
```
Comment (aidan-kwon) [inline blockchain/genesis_alloc.go:33]: After this change, `baobabGenesisJson` doesn't contain a full genesis. Currently, the value is used in a safe way. But, I am afraid that someone uses this value confusing with full genesis.

**Comment by yoomee1313** [comment]
@KlaytnDev As we implement incompatible change using code format, this change is no longer necessary. I will close this PR.

**Comment by yoomee1313** [comment]
@KlaytnDev Since this PR is derived for future blocknumber field addition, I thought this change is not necessary, because i'm not planning to add blocknumber. However, It was just stupid idea.. blocknumber is necessary, so i will reopen this PR after some fixes.

File: blockchain/genesis_alloc.go (hunk 4 above)
Comment (yoomee1313) [reply blockchain/genesis_alloc.go]: You're right.. I restored genesis_alloc.go chainconfig part.

File: params/config.go (hunk 3 above)
Comment (yoomee1313) [reply params/config.go]: Thanks, i changed the networkid! Could you please review again?

File: params/config.go (hunk 2 above)
Comment (yoomee1313) [reply params/config.go]: Thanks, i changed the networkid.

File: params/config.go (hunk 1 above)
Comment (yoomee1313) [reply params/config.go]: I forced push the commit.. Could you review again?

File: blockchain/genesis.go [hunk 5]
```diff
@@ -364,33 +364,28 @@ func GenesisBlockForTesting(db database.DBManager, addr common.Address, balance
 	return g.MustCommit(db)
 }
 
-// DefaultGenesisBlock returns the Klaytn main net genesis block.
+// DefaultGenesisBlock returns the Cypress mainnet genesis block.
+// It is also used for default genesis block.
 func DefaultGenesisBlock() *Genesis {
```
Comment (ehnuje) [inline blockchain/genesis.go:369]:
```suggestion
func DefaultCypressBlock() *Genesis {
```

File: params/config.go [hunk 6]
```diff
@@ -30,22 +30,63 @@ import (
 
 // Genesis hashes to enforce below configs on.
 var (
-	MainnetGenesisHash      = common.HexToHash("// todo generate new hash for mainnet") // Mainnet genesis hash to enforce below configs on
-	TestnetGenesisHash      = common.HexToHash("// todo generate new hash for testnet") // Testnet genesis hash to enforce below configs on
+	CypressGenesisHash      = common.HexToHash("// todo generate new hash for cypress") // cypress genesis hash to enforce below configs on
```
Comment (ehnuje) [inline params/config.go:33]: Is it okay to change the hash value here?

File: blockchain/genesis.go (hunk 5 above)
Comment (yoomee1313) [reply blockchain/genesis.go:369]: As mentioned in the comment, it is also used for default genesis block. So, I didn't changed the name.

File: params/config.go (hunk 6 above)
Comment (yoomee1313) [reply params/config.go:33]: @ehnuje Those variables are not actually being used. I just changed the name of it to match with.

**Comment by yoomee1313** [comment]
@ehnuje @winnie-byun @aidan-kwon I reflected your reviews. Please take a look. https://github.com/klaytn/klaytn/pull/882/commits/5a1f4d15069c71e253e92a06bca3b739457fa9ad.

File: params/config.go [hunk 7]
```diff
@@ -120,6 +161,8 @@ const (
 type ChainConfig struct {
 	ChainID *big.Int `json:"chainId"` // chainId identifies the current chain and is used for replay protection
 
+	Incompatible1Block *big.Int `json:"incompatible1Block"` // incompatible1 switch block (nil = no fork, 0 = already incompatible1)
```
Comment (ehnuje) [inline params/config.go:164]: This is a minor comment and a matter of taste, wouldn't it be better if the number is located at the end of the variable name for better readability? Like `IncompatibleBlock1`.

Comment (yoomee1313) [reply params/config.go:164]: @ehnuje Thanks for the comment, i didn't think about it.. However, i'm planning to change the name of Incompatible1 to IstanbulCompatible :). Are you okay with this name? It means a compatible with istanbul hardfork.

Comment (ehnuje) [reply params/config.go:164]: @yoomee1313 If so, `IstanbulCompatible` "only includes" the changes of Istanbul hardfork or "also includes" the changes of Klaytn hardfork?

Comment (yoomee1313) [reply params/config.go:164]: @ehnuje If this only includes Istanbul hardfork, then i wouldn't have named it IstanbulCompatible. In other words, Klaytn's hard fork items will also be included.

File: params/config.go [hunk 8]
```diff
@@ -30,22 +30,63 @@ import (
...
 
 var (
-	// MainnetChainConfig is the chain parameters to run a node on the main network.
-	MainnetChainConfig = &ChainConfig{
-		ChainID: big.NewInt(1),
-		Gxhash:  new(GxhashConfig),
+	// CypressChainConfig is the chain parameters to run a node on the cypress main network.
+	CypressChainConfig = &ChainConfig{
+		ChainID:            big.NewInt(int64(CypressNetworkId)),
+		Incompatible1Block: nil,
```
Comment (KimKyungup) [inline params/config.go:43]: This is next strong incompatable block number.

File: params/config.go (hunk 7 above)
Comment (ehnuje) [reply params/config.go:164]: @yoomee1313 Thanks for the explanation. I understood. 😄

**Review by ehnuje — APPROVED** [review]

**Review by KimKyungup — APPROVED** [review]

**Comment by yoomee1313** [comment]
Updated comment

**Comment by ehnuje** [comment]
Please check this change @jeongkyun-oh @aidan-kwon

File: params/config.go (hunk 7 above)
Comment (aidan-kwon) [inline params/config.go:164]: Doesn't it need `omitempty` tag?

**Review by aidan-kwon — APPROVED** [review]
LGTM except for a minor comment

**Review by jeongkyun-oh — APPROVED** [review]

File: params/config.go (hunk 7 above)
Comment (KimKyungup) [reply params/config.go:164]: @yoomee1313 What do you think?

Comment (yoomee1313) [reply params/config.go:164]: yes, it needs omitempty tag. I'll fix it in later change(PR) which changes the name of this field! @aidan-kwon @KimKyungup

Comment (aidan-kwon) [reply params/config.go:164]: Okay. Thanks!

**Pull request merged**